File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution 
{
    public:
        vector<int> postorderTraversal(TreeNode* root) 
        {
            vector<int> values;
            if (!root) return values; 

            vector<TreeNode*> stackRoot = {root};
            vector<TreeNode*> stackNode;   
        
            while (!stackRoot.empty())
            {
                auto node = stackRoot.back();

                stackNode.push_back(node);
                stackRoot.pop_back();

                if (node->left)
                    stackRoot.push_back(node->left);

                if (node->right)
                    stackRoot.push_back(node->right);
            }

            while (!stackNode.empty())
            {
                auto node = stackNode.back();
                values.push_back(node->val);
                stackNode.pop_back();
            }

            return values;
        }
};

File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution 
{
    public IList<int> PostorderTraversal(TreeNode root) 
    {
        List<int> values = new List<int>();

        if (root == null) return values;

        List<TreeNode> stackRoot = new List<TreeNode>() { root };
        List<TreeNode> stackNode = new List<TreeNode>();

        while (stackRoot.Count > 0)
        {
            var node = stackRoot[stackRoot.Count - 1];

            stackNode.Add(node);
            stackRoot.RemoveAt(stackRoot.Count - 1);

            if (node.left != null)
                stackRoot.Add(node.left);
            
            if (node.right != null)
                stackRoot.Add(node.right);
        }

        while (stackNode.Count > 0)
        {
            var node = stackNode[stackNode.Count - 1];

            values.Add(node.val);
            stackNode.RemoveAt(stackNode.Count - 1);
        }

        return values;
    }
}